# 分析目标

## 此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

# 简介

## 原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况。涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结场纪念品、圣诞礼品等等。该公司的客户群体生要包括批发商和个人消费者。其中批发商占据了相当大的比例。

## 数据每列的含义如下：
## InvoiceNo：发票号码。6位数,作为交易的通一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
## StockCode:产品代用。5位数。作为产品的唯一标识符。
## Description:产品名称。
## Quantity：产品在交易中的数量。
## InvoiceDate:发票日期和时间。交易发生的日期和时间。
## UnitPrice：单价。价格单位为英镑（£）。
## CustomerID:客户编号，5位数，作为客户的唯一标识符。
## Country:国家名称。客户所居住的国家的名称。

# 读取数据

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv("Online_Retail1.csv",encoding="utf-8")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  2012/1/10 8:26       2.55     17850.0  United Kingdom  
1  2012/1/10 8:26       3.39     17850.0  United Kingdom  
2  2012/1/10 8:26       2.75     17850.0  United Kingdom  
3  2012/1/10 8:26       3.39     17850.0  United Kingdom  
4  2012/1/10 8:26       3.39     17850.0  United Kingdom

# 评估数据

## 评估数据整齐度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
61138     541423     22324               BLUE POLKADOT KIDS BAG         1   
32508     539084     20972   PINK CREAM FELT CRAFT TRINKET BOX         12   
206465    554938     21884            CAKES AND BOWS GIFT  TAPE        12   
89567     543910     21984     PACK OF 12 PINK PAISLEY TISSUES         24   
174495    551820     22382           LUNCH BAG SPACEBOY DESIGN         10   
382028    569897     22252  BIRDCAGE DECORATION TEALIGHT HOLDER         2   
509383    579300     23215     JINGLE BELL HEART ANTIQUE SILVER        24   
106219    545317     21061               PARTY INVITES FOOTBALL         3   
459523    575875     22907     PACK OF 20 NAPKINS PANTRY DESIGN         5   
274838    560926     84991          60 TEATIME FAIRY CAKE CASES         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
61138      1/17/11 17:54       4.13         NaN  United Kingdom  
32508      12/16/10 9:31       1.25     13890.0  United Kingdom  
206465     5/27/11 13:34       0.65     12876.0         Belgium  
89567      2/14/11 12:36       0.29     13475.0  United Kingdom  
174495   2005/4/11 12:11       1.65     12645.0         Germany  
382028   2010/6/11 16:01       1.25     17813.0  United Kingdom  
509383    11/29/11 11:30       2.08     14667.0  United Kingdom  
106219   2003/1/11 14:14       1.63         NaN  United Kingdom  
459523  2011/11/11 13:06       1.63         NaN  United Kingdom  
274838      7/22/11 9:20       1.25         NaN  United Kingdom

## 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


### 从输出结果来看。数据共有541909条观察值，而Description、CustomerID变量存在缺失值。此外，InvoiceDate的数据类型应为日期，CustomerID的数据类型应为字符串，应当进行数据格式转换。

## 评估缺失数据

### 根据条件提取出缺失的观察值

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  2012/1/10 11:52        0.0   
1970      536545     21134         NaN         1  2012/1/10 14:32        0.0   
1971      536546     22145         NaN         1  2012/1/10 14:33        0.0   
1972      536547     37509         NaN         1  2012/1/10 14:33        0.0   
1987      536549    85226A         NaN         1  2012/1/10 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  2012/7/11 18:26        0.0   
535326    581203     23406         NaN        15  2012/7/11 18:31        0.0   
535332    581209     21620         NaN         6  2012/7/11 18:35        0.0   
536981    581234     72817         NaN        27  2012/8/11 10:33        0.0   
538554    581408     85175         NaN        20  2012/8/11 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

#### 发现单价好像都为0，为了验证猜想看是不是

In [7]:
original_data[(original_data["Description"].isnull())&(original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### 猜想成功，这些无效的东西后续可以删除

#### CustomerID同样有确实所以看看

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     2012/1/10 11:52       0.00         NaN  United Kingdom  
1443    2012/1/10 14:32       2.51         NaN  United Kingdom  
1444    2012/1/10 14:32       2.51         NaN  United Kingdom  
1445    2012/1/10 14:32       0.85         NaN  United Kingdom  
1446    2012/1/10 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  2012/9/11 10:26       4.13         NaN  United Kingdom  
541537  2012/9/11 10:26       4.13         NaN  United Kingdom  
541538  2012/9/11 10:26       4.96         NaN  United Kingdom  
541539  2012/9/11 10:26      10.79         NaN  United Kingdom  
541540  2012/9/11 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

#### 虽然ID无效，但其他数据有用，不删

### 评估重复数据

#### 这个数据集不需要考虑

### 评估不一致数据

#### 有可能数据表示同一个信息但是值不一样，比如此数据集的Country,所以要进行分析

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


### 评估无效或错误数据

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

#### 发现质量和单价有负数的情况出现->检查

In [11]:
original_data[original_data["Quantity"]<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      2012/1/10 9:41      27.50     14527.0  United Kingdom  
154      2012/1/10 9:49       4.65     15311.0  United Kingdom  
235     2012/1/10 10:24       1.65     17548.0  United Kingdom  
236     2012/1/10 10:24       0.29     17548.0  United Kingdom  
237     2012/1/10 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   2012/9/11 9:57       0.83     14397.0  United Kingdom  
541541  2012/9/11 10:28     224.69     15498.0  United Kingdom  
541715  2012/9/11 11:57      10.95     15311.0  United Kingdom  
541716  2012/9/11 11:58       1.25     17315.0  United Kingdom  
541717  2012/9/11 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [12]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  2012/1/10 16:50        0.0   
4347      536764    84952C         NaN       -38  2012/2/10 14:42        0.0   
7188      536996     22712         NaN       -20  2012/3/10 15:30        0.0   
7189      536997     22028         NaN       -20  2012/3/10 15:30        0.0   
7190      536998     85067         NaN        -6  2012/3/10 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  2012/7/11 18:36        0.0   
535335    581212     22578        lost     -1050  2012/7/11 18:38        0.0   
535336    581213     22576       check       -30  2012/7/11 18:38        0.0   
536908    581226     23090     missing      -338   2012/8/11 9:56        0.0   
538919    581422     23169     smashed      -235  2012/8/11 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [13]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")&(original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### 根据输出结果。当Quantity变量为负数的时候，观察值满足以下条件之一：InvoiceNo以C开头。表示订单被取消UnitPrice为0，说明单价为0英镑这些交易数据均不是有效成交数据。不贡献销售。删

In [14]:
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity       InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  2008/12/11 14:51   
299984   A563187         B  Adjust bad debt         1  2008/12/11 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

## 清理数据

需要进行的操作有：
- InvoiceDate的数据变量类型改成日期时间
- CustomerID的数据变量类型改为字符串
- Description变量缺失的观察值删除
- Quantity为负数的删掉
- UnitPrice为负数的删掉

In [15]:
cleaned_data=original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  2012/1/10 8:26       2.55     17850.0  United Kingdom  
1  2012/1/10 8:26       3.39     17850.0  United Kingdom  
2  2012/1/10 8:26       2.75     17850.0  United Kingdom  
3  2012/1/10 8:26       3.39     17850.0  United Kingdom  
4  2012/1/10 8:26       3.39     17850.0  United Kingdom

In [16]:
cleaned_data["InvoiceDate"]=pd.to_datetime(cleaned_data["InvoiceDate"],format='mixed')
cleaned_data["InvoiceDate"]

0        2012-01-10 08:26:00
1        2012-01-10 08:26:00
2        2012-01-10 08:26:00
3        2012-01-10 08:26:00
4        2012-01-10 08:26:00
                 ...        
541904   2012-09-11 12:50:00
541905   2012-09-11 12:50:00
541906   2012-09-11 12:50:00
541907   2012-09-11 12:50:00
541908   2012-09-11 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html
可以查找这个的用法

In [17]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [18]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].str.slice(0,5)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

或者写slice(0,-2)，这里ID肯定不是有小数的所以后的要删掉

In [21]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [23]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

In [24]:
cleaned_data=cleaned_data[cleaned_data["Quantity"]>=0]

最好检查一下

In [25]:
cleaned_data=cleaned_data[cleaned_data["UnitPrice"]>=0]

最好检查一下

## 保存清理好的数据

In [26]:
cleaned_data.to_csv("Online_Retail1_cleaned.csv")